coveo_stew/ci/runner.py

```diff
@@ -119,3 +137,139 @@ def _output_generic_report(self, environment: PythonEnvironment) -> None:
 
     def __str__(self) -> str:
         return self.name
+
+
+@dataclass
+class CIPlan:
+    """
+    The CIPlan holds all the runners for an environment
+    and orchestrates the workflow of launching them.
+    """
+
+    environment: PythonEnvironment
+    checks: Sequence[ContinuousIntegrationRunner]
+    parallel: bool
+
+    @cached_property
+    def autofix_checks(self) -> List[ContinuousIntegrationRunner]:
+        """
+        Autofix checks receive a special treatment since they will potentially change line numbers,
+        making reports inaccurate.
+        """
+        return [check for check in self.checks if check.supports_auto_fix]
+
+    @cached_property
+    def non_autofix_checks(self) -> List[ContinuousIntegrationRunner]:
+        return [check for check in self.checks if check not in self.autofix_checks]
+
+    async def orchestrate(self, auto_fix: bool = False) -> None:
+        """Orchestrates this CIPlan by launching runners in the correct order."""
+        runs: List[Run] = []
+        echo.step(
+            f"Planned {len(self.checks)} runners for {self.environment.pretty_python_version}"
+        )
+
+        if auto_fix:
+            # run autofix first, in parallel
+            runs.append(run := Run(self.environment, self.autofix_checks))
+            await run.run_and_report(parallel=self.parallel)
+
+            if run.overall_status is RunnerStatus.CheckFailed:
+                await run.run_and_report(auto_fix=True, feedback=False, parallel=False)
```

Comment:
Why do you need

![image](https://user-images.githubusercontent.com/8481892/173455300-b1a5c555-69df-4443-844c-29581086d9b5.png)

if you call the function with `parallel=False` here?

Reply:
just making sure I don't make mistakes in the future! :P

Reply:
i kept it in for precautions, but since that would hide a dev mistake (i.e. it's not possible for the user to do cause this) I changed it into an assert.